pkg/database/database.go

```diff
@@ -14,238 +15,167 @@ import (
 	if err != nil {
 		return fmt.Errorf("cannot get query by name for namespace %s: %w", namespace, err)
 	}
-	ctx.query = queryMatch
+	dctx.query = queryMatch
 
 	return nil
 }
 
-func (d *Database) adjustStorageSizeUnit(ctx *transactionContext) error {
-	var quantity float64
-	if query.Unit == defaultUnit {
-		quantity = ctx.aggregatedBucket.StorageUsed / 1024 / 1024 / 1024
-	} else {
-		return fmt.Errorf("unknown query unit %s", query.Unit)
+func (d *Database) saveFacts(dctx *Context) error {
+	storageFact := factsmodel.New(dctx.dateTime, dctx.query, dctx.tenant, dctx.category, dctx.product, dctx.discount, *dctx.value)
+	_, err := factsmodel.Ensure(dctx, dctx.transaction, storageFact)
+	if err != nil {
+		return fmt.Errorf("cannot save fact for namespace %s: %w", *dctx.namespace, err)
 	}
-	ctx.quantity = &quantity
+
 	return nil
 }
 
-func (d *Database) saveFacts(ctx *transactionContext) error {
-	storageFact := factsmodel.New(ctx.dateTime, ctx.query, ctx.tenant, ctx.category, ctx.product, ctx.discount, *ctx.quantity)
-	_, err := factsmodel.Ensure(ctx, ctx.transaction, storageFact)
+// commitTransaction commits a transaction in the billing database
+func (d *Database) commitTransaction(dctx *Context) error {
+	err := dctx.transaction.Commit()
 	if err != nil {
-		return fmt.Errorf("cannot save fact for namespace %s: %w", *ctx.namespace, err)
+		return fmt.Errorf("cannot commit transaction in the database: %w", err)
 	}
 	return nil
 }
 
-func (d *Database) commitTransaction(ctx *transactionContext) error {
-	err := ctx.transaction.Commit()
+// beginTransaction creates a new transaction in the billing database
+func (d *Database) beginTransaction(dctx *Context) error {
+	tx, err := d.connection.BeginTxx(dctx, &sql.TxOptions{})
 	if err != nil {
-		return fmt.Errorf("cannot commit transaction for buckets in namespace %s: %w", *ctx.namespace, err)
+		return fmt.Errorf("cannot create database transaction: %w", err)
 	}
+	dctx.transaction = tx
 	return nil
 }
 
-// EnsureInitConfiguration ensures the minimum exoscale object storage configuration data is present in the database
-// before saving buckets usage
-func (d *Database) EnsureInitConfiguration(ctx context.Context) error {
-	transaction, err := d.connection.BeginTxx(ctx, &sql.TxOptions{})
-	if err != nil {
-		return fmt.Errorf("cannot begin transaction for initial database configuration: %w", err)
-	}
-	defer transaction.Rollback()
-	err = ensureInitConfigurationModels(ctx, err, transaction)
-	if err != nil {
-		return err
-	}
-	err = transaction.Commit()
+// rollback rolls back transaction in case of error in previous steps
+func (d *Database) rollback(dctx *Context, err error) error {
+	log := ctrl.LoggerFrom(dctx)
 	if err != nil {
-		return fmt.Errorf("cannot commit transaction for initial database configuration: %w", err)
+		log.Error(err, "found error in pipeline")
+		err = dctx.transaction.Rollback()
+		if err != nil {
+			return fmt.Errorf("cannot rollback transaction: %w", err)
+		}
```

Comment:
if there was an error and the rollback was successful, this would return no error and possibly exit code 0. Is that intentional?

Reply:
I think it's a potential bug here yes. This method is being used not only for aggregated objects but for ensuring the initial billing database configuration. At this point we might stop the application altogether even for single aggregate objects. We might miss as you said object not being saved in DB and no one would notice it in the logs.